### PR TITLE
Deal with situation that no SimpleCredentials are set

### PR DESCRIPTION
**Pull request by holtkamp**

### CODE REVIEW

**Review by dbu — COMMENTED** [review]
thank you, agreed that this is implied to work. we probably broke it at some later point by adding the NodeProcessor...

i am not sure if empty string is the right thing, or we should use something like 'unknown' or if we should not set the properties in that case.

and can you please add a test to ClientTest that tests the login with `null` and then does something that triggers getNodeProcessor, e.g. by calling updateProperties?

**Comment by holtkamp** [comment]
Yeah, will try to spend some more time on this coming week. Needed a patch quickly to get rid of the error.

File: src/Jackalope/Transport/DoctrineDBAL/Client.php [hunk 1]
```diff
@@ -2017,7 +2017,7 @@ private function getNodeProcessor()
         }
 
         $this->nodeProcessor = new NodeProcessor(
-            $this->credentials->getUserID(),
+            $this->credentials ? $this->credentials->getUserID() : '',
```
[the remark this answers is not in the document]
Comment (holtkamp) [reply src/Jackalope/Transport/DoctrineDBAL/Client.php]: personally I would go for the third option: use `null`.

**Comment by holtkamp** [comment]
had a quick look at the specification of JSR-283

http://download.oracle.com/otndocs/jcp/content_repository-2.0-fr-oth-JSpec/

Unfortunately no XML Schema is included. @dbu does this exist?

As I understand it, the `jcr:createdBy` and `jcr:lastModifiedBy` are variant attributes which means:

```
3.7.9.1 Implementation Variants in Node Types
Some of the attributes of the node types defined in this specification may vary across implementations. For example, it is implementation-dependent which node types and which properties are queryable (see §3.7.1.5 Queryable Node and §3.7.3.3 Available Query Operators). Similarly, some of the standard application node types (see §3.7.11 Standard Application Node Types) may vary as to the on-parent-version and protected status of some properties.
```

So I guess the question is, are these attributes "queryable" in the PHP implementation of JSR-283? If they are, the fields should be present but can be `null`. Personally I use the following interpretation when dealing with XML:

- `<element></element>` => empty string: ''
- `<element/>` => `null`

Unfortunately no strict rules exist for this interpretation in XML. But for a `null` value this would mean:

```xml
<sv:property sv:name="jcr:createdBy" sv:type="String" sv:multi-valued="0">
    <sv:value/>
</sv:property>
```

Currently an empty string '' is encoded by PHPCR like:
```xml
<sv:property sv:name="jcr:createdBy" sv:type="String" sv:multi-valued="0">
    <sv:value length="0"></sv:value>
</sv:property>
```

Are XML Schemas available? This might help in making the proper decision. For example is the `length` attribute required on an empty `value` element?

**Comment by dbu** [comment]
3.7.11.7 in https://docs.adobe.com/docs/en/spec/jcr/2.0/3_Repository_Model.html says that the createdBy SHOULD be populated. so i guess it would be ok to not set it no information is available. i don't know if there are any side effects. 

until now, things simply failed when no credentials have been provided. i would say lets use `null` and then in the processor simply skip the createdBy and lastUpdatedBy properties, and add a test in jackalope-doctrine-dbal with `null`credentials that creates and fetches a node to see if it works like that.

the processor is in the jackalope/jackalope repository, so it needs its own pull request. can you please make this PR then use the branch of your PR to jackalope/jackalope so that we can see the whole thing working before merging jackalope/jackalope?

**Comment by holtkamp** [comment]
It seems I only just now saw your detailed comment: https://github.com/jackalope/jackalope-doctrine-dbal/pull/359#issuecomment-392246788 and I am not really sure I understand what's the idea:

- open a PR at https://github.com/jackalope/jackalope to make sure the [NodeProcessor](https://github.com/jackalope/jackalope/blob/c1ea301cd12b10b7cd70b0ff6ee2c1cbb71deb50/src/Jackalope/NodeType/NodeProcessor.php) skips `createdBy` and `lastUpdatedBy` (should be `lastModifiedBy`?) properties: https://github.com/jackalope/jackalope/blob/c1ea301cd12b10b7cd70b0ff6ee2c1cbb71deb50/src/Jackalope/NodeType/NodeProcessor.php#L190-L191 ??

**Comment by dbu** [comment]
oh hi, old issue that one. i just read up on our discussion again. i think handling null in the NodeProcessor is the way to go, yes. (and sorry for the mixup, i meant to say lastModifiedBy and not lastUpdatedBy). we should also change the phpdoc on $userId there to say it is `string|null` and in constructor not cast null to string (but still cast other things to string for BC).

we also need to adjust https://github.com/jackalope/jackalope/blob/c1ea301cd12b10b7cd70b0ff6ee2c1cbb71deb50/src/Jackalope/NodeType/NodeProcessor.php#L235 - i think the correct thing to do there would be to remove the lastModifiedBy property in that case to not lie. (i assume in normal situations, you either always or never have the userid, so we probably don't find an existing lastModifiedBy when we don't have the user id. but in case we do, it seems better than to skip updating and lie about who did the last modified)

**Comment by dbu** [comment]
i can then tag a patch version of jackalope, and you can make this repo require at least that patch version and pass null.
